### ISSUE TITLE
Research:  Improve the ranking of the sources

### ISSUE THREAD

**Issue by vatsrahul1001** [astronomer/ask-astro#133]
While Testing Ask Astro today we noticed some issue with below questions

1. what is Astro SDK?
   The response for this was incorrect(I'm sorry, but there's no such thing as Astro-SDK.) and the sources were irrelevant.
    [Slack Thread ](https://ankit-test-group.slack.com/archives/C0612F9CB18/p1700146230528799)

2. What is latest version for Astronomer Providers
    The response was not correct as per Ask Astro latest version is 1.14.0, however, latest release is 1.18.2
   [ Slack Thread](https://ankit-test-group.slack.com/archives/C061F4LR4KT/p1700145976195339?thread_ts=1700145891.020609&cid=C061F4LR4KT)

3. Related docs incorrect for how to install the CLI on Linux?
   [Slack Thread](https://ankit-test-group.slack.com/archives/C061F4LR4KT/p1700148903485149?thread_ts=1700147849.918279&cid=C061F4LR4KT)

4. Ask Astro not recognize version 2.7.3 and treat 2.2.2 as latest
[   Slack Thread](https://ankit-test-group.slack.com/archives/C061F4LR4KT/p1700224963253469)

Based on above try the following in the order: 
1. tokenization as lowercase [#145](https://github.com/astronomer/ask-astro/issues/145)
2. Cohere reranking - [PR](https://github.com/astronomer/ask-astro/pull/170)
3. hybrid search - [PR](https://github.com/astronomer/ask-astro/pull/170)

Test iteration for each experiment by @vatsrahul1001 

#### --- Update 1/8 Merging two other similar/duplicate issues to this one, closing the other two ---
https://github.com/astronomer/ask-astro/issues/213
https://github.com/astronomer/ask-astro/issues/80

**Comment by sunank200** [comment]
Here is Langsmith trace for 
1. Yes, for `what is the Astro SDK?` The following is langsmith [trace](https://smith.langchain.com/o/4942ae8b-e5be-4964-bb1f-7544886fa06d/projects/p/5d1aeb88-dc89-4f37-9e7c-3f5a3cf973e2?timeModel=%7B%22start_time%22%3A%222023-11-13T15%3A16%3A03.451Z%22%7D&peek=5b17b072-e838-41fa-88e9-e8d16adb8a25). It doesn't find the correct docs in Retriever. But it should have either picked [[1]](https://www.astronomer.io/blog/migrate-python-jobs-to-airflow-in-4-simple-steps/), [[2]](https://github.com/astronomer/docs/blob/main/learn/astro-python-sdk-etl.md), [[3]](https://www.astronomer.io/blog/ask-astro-open-source-llm-application-apache-airflow/), [[4]](https://github.com/astronomer/docs/blob/main/astro/features.md), [[5]](https://github.com/astronomer/docs/blob/main/learn/astro-python-sdk.md) which is already ingested.

But when we ask -  `What is Astro Python SDK?` It gives correct answers with the right sources. Here is langsmith [trace](https://smith.langchain.com/o/4942ae8b-e5be-4964-bb1f-7544886fa06d/projects/p/5d1aeb88-dc89-4f37-9e7c-3f5a3cf973e2?timeModel=%7B%22start_time%22%3A%222023-11-13T15%3A16%3A03.451Z%22%7D&peek=e6ef28cf-dea4-42a5-98ba-4d6ba2f4783d)

**Comment by sunank200** [comment]
- For thie following question:
```
Hi all ,

      Need your help and guidance

      I am going to install airflow version 2.7.3 in new Ec2 instance with postgres

      EC2 - t3.x large (4 vCPUs, 16 GB RAM )

      What is the ubuntu and python version that will be compatible with this ?

      Anyone who installed 2.7.3 can share your thought
```
Langsmith trace is [here](https://smith.langchain.com/o/4942ae8b-e5be-4964-bb1f-7544886fa06d/projects/p/5d1aeb88-dc89-4f37-9e7c-3f5a3cf973e2?timeModel=%7B%22start_time%22%3A%222023-11-13T15%3A16%3A03.451Z%22%7D&peek=9a68840e-9f65-4f80-b6ac-af19c110f3d8)

For `What is the latest Airflow version?` it gave the correct response though. Langchain trace is [here](https://smith.langchain.com/o/4942ae8b-e5be-4964-bb1f-7544886fa06d/projects/p/5d1aeb88-dc89-4f37-9e7c-3f5a3cf973e2?timeModel=%7B%22start_time%22%3A%222023-11-13T15%3A16%3A03.451Z%22%7D&peek=f103f49d-2c3a-452b-bcff-634e4a1a9647)

**Comment by mpgreg** [comment]
Initial analysis is at https://docs.google.com/document/d/17OBh5b9fQM3kq_n1fxbIL49b2fM0-ipPPD-Ju0_2nIo/edit

**TLDR;** 
The docSource property is vectorized during ingest and search.  This is skewing search results containing ‘astro’ and ‘astronomer’  towards certain sources.  
By itself the “source skew”  is not a problem but without hybrid search the vector will skew somewhat randomly towards.
Formatting of Astro docs ingested from Markdown confuses the LLM for answering.

**Recommendations:**
Remove docSource.  It is not used and inconsistently named.
Or… change schema for docSource to skip vectorization.
Implement hybrid search
Change astro doc ingest to extract from HTML sources

**Comment by mpgreg** [comment]
I'm testing now with local docs with schema change for ```"properties": [
                {
                    "name": "docSource",
                    "description": "Type of document ('learn', 'astro', 'airflow', 'stackoverflow', 'code_samples')",
                    "dataType": ["text"],
                    "moduleConfig": {
                        "text2vec-openai": {
                            "skip": "**True**",
                            "vectorizePropertyName": "False"
                        }
                    }
                },```

**Comment by sunank200** [comment]
We tried [hybrid search with Cohere reranking](https://github.com/astronomer/ask-astro/pull/170) and this has degraded the performance. Hence not a priority for 28th Nov release

**Comment by phanikumv** [comment]
David to look into this

**Comment by sunank200** [comment]
David to have first results by EOW

**Comment by sunank200** [comment]
@davidgxue any updates on this?

**Comment by davidgxue** [comment]
My apologies I forgot to update on Github. I sync’ed with Steven on Friday and send out a google doc that contains approaches to experiment with based on initial analysis and obviously (https://docs.google.com/document/d/1j-Hr8dchwBWDxejAf1dvcGIA_Y6lVQ-W9VP0k7Zw9zE/edit?usp=sharing). I am currently on PTO this entire week but I will update with more details when I get back.

**Comment by davidgxue** [comment]
## Update: Research Report

### Next Steps and Approaches for Experimentation and Implementation

#### Retrieval Enhancements
1. **Hybrid / Sparse + Dense Vector Search Integration with Cohere Reranking**
   - Implement a hybrid search combining BM25 and an embedding model with rank fusion scoring to narrow down results to the top 100-300 documents.
   - Rerank the shortlisted documents using Cohere to refine the selection to approximately 10 documents.
   - This approach aims to improve performance and reduce latency compared to the sequential use of BM25, embedding models, and rerankers.

2. **Language Model Prompt Rewording for Multi-Query Retrieval**
   - Maintain the original user prompt as one of the queries to preserve the initial intent.
   - Optimize the rewording prompt for GPT-3.5 to ensure it is concise and summarizes the query without introducing extraneous content.

3. **Parent Document Retriever Implementation**
   - Address the issue where relevant keywords or related terms appear in one section of a page, but the actual answers are located in a different section of the same document.

4. **Final Relevance Check with a Cost-Effective LLM**
   - Utilize a less resource-intensive language model like GPT-3.5 (e.g., LLMChainFilter in LangChain) to assess the relevance of the retrieved documents before processing them with GPT-4 for final response generation.

#### Response Generation Improvements
1. **Prompt Engineering for Source Citation**
   - Refine the system prompt that triggers GPT-4 for final response generation to require explicit citation of the source document used in the answer, as recommended by Julian. This should prevent the generation of responses without proper backing from source documents.

#### Data Management and Optimization

##### Vector Database (Vector DB)
1. **Exploration of Alternative Embedding Models**
   - Investigate the use of embedding models from Cohere as potential alternatives to the `text-ada-002` model from OpenAI.

2. **Vector DB Schema Modification**
   - Adjust the Vector DB schema to exclude the vectorization of certain non-essential attributes, such as `docSource`.

##### Data Ingestion Process
1. **Data Cleaning and Exclusion of Irrelevant Content**
   - Implement data cleaning during ingestion to remove non-essential content like navigation bars, footers, headers, and other irrelevant sections that may introduce keyword spam and reduce retrieval accuracy.

2. **Review and Refinement of Chunking Logic**
   - Reassess the logic for document chunking to prevent the inclusion of headers or short, meaningless text segments.

3. **Summarization of Large Documents**
   - Generate and insert summaries for excessively large documents that are split into numerous chunks, using a language model to aid in comprehension and retrieval.

4. **Topic Keyword Extraction and Metadata Storage**
   - Perform topic keyword extraction on each document in the Vector DB, store the results as metadata, and enhance queries with user-prompt-derived keywords during Q&A sessions. This strategy requires significant effort and its effectiveness is yet to be determined.

### Current Progress Update

- The enhancements listed under **Retrieval** items 1, 2, and 4 have been implemented as part of this pull request: [PR #247](https://github.com/astronomer/ask-astro/pull/247). This PR will resolve the associated issue for the current sprint.
- To address the issue of noisy data from **Data Ingestion** item 1, the most problematic document source containing excessive navbar, footer, and header content has been temporarily excluded from the ingestion process.
- Exploration and development of the remaining items are ongoing.
